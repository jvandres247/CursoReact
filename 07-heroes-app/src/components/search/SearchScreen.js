import React, { useMemo } from "react";
import { useLocation } from "react-router-dom";
import { heroes } from "../../data/heroes";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroCard";
import queryString from "query-string";
import { getHeroesByName } from "../../selectors/getHeroesByName";

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const [{ searchText }, handleInputChange, reset] = useForm({
    searchText: q,
  });
  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${searchText}`);
  };

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);
  return (
    <div>
      <h1>Search Screen</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Search form</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Find your hero"
              className="form-control"
              autoComplete="off"
              name="searchText"
              value={searchText}
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === "" && <div className="alert alert-info">Search a Hero...</div>}
          {q !== "" && heroesFiltered.length === 0 && (
            <div className="alert alert-danger">
              There is no a hero with {q}
            </div>
          )}
          {heroesFiltered.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};
