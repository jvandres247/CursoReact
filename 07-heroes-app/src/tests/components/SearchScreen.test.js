import React from "react";
import { mount } from "enzyme";
import { SearchScreen } from "../../components/search/SearchScreen";
import { MemoryRouter, Route } from "react-router-dom";

describe("Pruebas en <SearchScreen/>", () => {
  test("Debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe("Search a Hero...");
  });

  test("Debe de mostrar a Batman y el input con el valor del query string", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman123"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    //expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      "There is no a hero with batman123"
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de llamar el push del history", () => {
    const history = {
        push: jest.fn(),
        goBack: jest.fn(),
      };

      const wrapper = mount(
        <MemoryRouter initialEntries={["/search?q=batman"]}>
          <Route path="/search" component={()=> <SearchScreen history={history}/>} />
        </MemoryRouter>
      );

      wrapper.find('input').simulate('change', {
          target: {
              name: 'searchText',
              value: 'batman'
          }
      });

      wrapper.find('form').prop('onSubmit')({
          preventDefault(){}
      });

      expect(history.push).toHaveBeenCalledWith('?q=batman');
    
  });
});
