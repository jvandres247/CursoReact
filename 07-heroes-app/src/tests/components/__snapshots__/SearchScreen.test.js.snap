// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen/> Debe de mostrar a Batman y el input con el valor del query string 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        There is no a hero with 
        batman123
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen/> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero...
      </div>
    </div>
  </div>
</div>
`;
